File: js/euler61.js
var triangles = new Set();
var squares = new Set();
var pentagons = new Set();
var hexagons = new Set();
var heptagons = new Set();
var octagons = new Set();
var polygonSets = [triangles, squares, pentagons, hexagons, heptagons];

function euler61(){
    // find 4-digit figurate numbers
    for (let n = 1000; n < 10_000; n++){
        if (isTriangle(n)) triangles.add(n);
        if (isSquare(n)) squares.add(n);
        if (isPentagon(n)) pentagons.add(n);
        if (isHexagon(n)) hexagons.add(n);
        if (isHeptagon(n)) heptagons.add(n);
        if (isOctagon(n)) octagons.add(n);
    }
    // cycle can be started from any polygon set
    // start with octagons, as it has the fewest candidates
    for (let octagon of octagons){
        let cycle = generateCycle([octagon], octagon, new Set());
        // calculate sum of completed cycle of 6 numbers
        if (cycle && cycle.length == 6) return cycle.reduce((total, n) => total + n, 0);
    }
}

function generateCycle(cycle, last, used){
    // full cycle of 6 numbers completed
    // last 2 digits of last number is first 2 digits of first number
    if (cycle.length == 6 && last % 100 == Math.floor(cycle[0] / 100)) return cycle;
    for (let s of polygonSets){
        // each polygon set is represented by a different number, may only be used once
        if (used.has(s)) continue;
        used.add(s);
        for (let p of s){
            // cycle can be continued with p
            // 
            if (last % 100 == Math.floor(p / 100)){
                cycle.push(p);
                // if result is not undefined, complete cycle has been found
                if (generateCycle(cycle, p, used)) return cycle;
                // backtrack
                cycle.pop();
            }
        }
        used.delete(s);
    }
}

function isTriangle(p){
    // p = n (n + 1) / 2
    // 0 = n^2 + n - 2p
    // n = (-1 + sqrt(1 + 8p)) / 2
    var n = (Math.sqrt(8*p + 1) - 1) / 2;
    return n == Math.floor(n);
}

function isSquare(p){
    var n = Math.sqrt(p);
    return n == Math.floor(n);
}

function isPentagon(p){
    // p = n * (3n - 1) / 2
    // 0 = 3n^2 - n - 2p
    // n = (1 + sqrt(1 + 24p)) / 6
    var n = (Math.sqrt(24*p + 1) + 1) / 2;
    return n == Math.floor(n);
}

function isHexagon(p){
    // p = n * (2n - 1)
    // 0 = 2n^2 - n - p
    // n = (1 + sqrt(1 + 8p)) / 4
    var n = (Math.sqrt(8*p + 1) + 1) / 4;
    return n == Math.floor(n);
}

function isHeptagon(p){
    // p = n * (5n - 3) / 2
    // 0 = 5n^2 - 3n - 2p
    // n = (3 + sqrt(9 + 40p)) / 10
    var n = (Math.sqrt(40*p + 9) + 3) / 10;
    return n == Math.floor(n);
}
function isOctagon(p){
    // p = n * (3n - 2)
    // 0 = 3n^2 - 2n - p
    // n = (2 + sqrt(4 + 12p)) / 6
    // n = (1 + sqrt(1 + 3p)) / 3
    var n = (Math.sqrt(3*p + 1) + 1) / 3;
    return n == Math.floor(n);
}

console.log(euler61());

File: go/euler61.go
package main
import (
    "fmt"
    "math"
)

func isTriangle(n int) bool {
    // p = n * (n + 1) / 2
    // 0 = n^2 + n - 2p
    // n = (-1 + sqrt(1 + 8p)) / 2
    x := (math.Sqrt(8 * float64(n) + 1) - 1) / 2
    return x == math.Floor(x)
}

func isSquare(n int) bool {
    x := math.Sqrt(float64(n))
    return x == math.Floor(x)
}

func isPentagon(n int) bool {
    // p = n * (3n - 1) / 2
    // 0 = 3n^2 - n - 2p
    // n = (1 + sqrt(1 + 24p)) / 6
    x := (math.Sqrt(24 * float64(n) + 1) + 1) / 6
    return x == math.Floor(x)
}

func isHexagon(n int) bool {
    // p = n * (2n - 1)
    // 0 = 2n^2 - n - p
    // n = (1 + sqrt(1 + 8p)) / 4
    x := (math.Sqrt(8 * float64(n) + 1) + 1) / 4
    return x == math.Floor(x)
}


func isHeptagon(n int) bool {
    // p = n * (5n - 3) / 2
    // 0 = 5n^2 - 3n - 2p
    // n = (3 + sqrt(9 + 40p)) / 10
    x := (math.Sqrt(40 * float64(n) + 9) + 3) / 10
    return x == math.Floor(x)
}

func isOctagon(n int) bool {
    // p = n * (3n - 2)
    // 0 = 3n^2 - 2n - p
    // n = (2 + sqrt(4 + 12p)) / 6
    // n = (1 + sqrt(1 + 3p)) / 3
    x :=(math.Sqrt(3 * float64(n) + 1) + 1) / 3
    return x == math.Floor(x)
}

func canContinue(a int, b int) bool {
    return a % 100 == b / 100
}

func generateCycle(cycle []int, used map[int]bool, polygonSets [][]int) []int {
    if len(cycle) == 6 {
        if canContinue(cycle[5], cycle[0]) {
            return cycle
        }
        return []int{}
    }
    last := cycle[len(cycle) - 1]
    for i, pSet := range polygonSets {
        if used[i] {
            continue
        }
        for _, p := range pSet {
            if canContinue(last, p) {
                used[i] = true
                newCycle := generateCycle(append(cycle, p), used, polygonSets)
                if len(newCycle) != 0 {
                    return newCycle
                }
                used[i] = false
            }
        }
    }
    return []int{}
}

func main() {
    triangles := []int{}
    squares := []int{}
    pentagons := []int{}
    hexagons := []int{}
    heptagons := []int{}
    octagons := []int{}
    
    for n := 1000; n < 10000; n++ {
        if isTriangle(n) {
            triangles = append(triangles, n)
        }
        if isSquare(n) {
            squares = append(squares, n)
        }
        if isPentagon(n) {
            pentagons = append(pentagons, n)
        }
        if isHexagon(n) {
            hexagons = append(hexagons, n)
        }
        if isHeptagon(n) {
            heptagons = append(heptagons, n)
        }
        if isOctagon(n) {
            octagons = append(octagons, n)
        }
    }
    polygons := [][]int{
        triangles, squares, pentagons, hexagons, heptagons,
    }
    res := 0
    
    // start with smallest set
    for _, oct := range octagons {
        cycle := generateCycle([]int{oct}, make(map[int]bool), polygons)
        
        if len(cycle) == 6 {
            sum := 0
            for _, p := range cycle {
                sum += p
            }
            res = sum
            break
        }
    }
    fmt.Println(res)
}
